### ISSUE TITLE
Need a story about implementing async-read, async-write

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/wg-async-foundations#181]
### Brief summary

I've heard that it's common want to wrap async-read or async-write, particularly with folks come from Java where this kind of pattern is very complicated. But it's hard to do because `Pin` (similar perhaps to "Alan Hates Writing a Stream")

### Optional details

* (Optional) Which [character(s)] would be the best fit and why?
    * [x] [Alan]: the experienced "GC'd language" developer, new to Rust -- maybe?
    * [ ] [Grace]: the systems programming expert, new to Rust
    * [ ] [Niklaus]: new programmer from an unconventional background
    * [ ] [Barbara]: the experienced Rust developer
* (Optional) Which [project(s)] would be the best fit and why?
    * *List some projects here.*
* (Optional) What are the key points or morals to emphasize?
    * *Write some morals here.*

[character(s)]: https://rust-lang.github.io/wg-async-foundations/vision/characters.html
[project(s)]: https://rust-lang.github.io/wg-async-foundations/vision/projects.html
[Alan]: https://rust-lang.github.io/wg-async-foundations/vision/characters/alan.html
[Grace]: https://rust-lang.github.io/wg-async-foundations/vision/characters/grace.html
[Niklaus]: https://rust-lang.github.io/wg-async-foundations/vision/characters/niklaus.html
[Barbara]: https://rust-lang.github.io/wg-async-foundations/vision/characters/barbara.html

**Comment by rylev** [comment]
FYI: We had a writing session on this today, and we should be having the first (of hopefully 2-3) status-quo stories coming for this soon.

**Comment by nikomatsakis** [comment]
ooh, nice!

**Comment by eminence** [comment]
An interesting aspect to this story could be:   does Alan write an impl for tokio's [AsyncWrite](https://docs.rs/tokio-io/0.1.13/tokio_io/trait.AsyncWrite.html) or future's [AsyncWrite](https://docs.rs/futures/0.3.14/futures/io/trait.AsyncWrite.html) ?

**Comment by jbr** [comment]
I'm not sure if "wrap" in the description is the same as this, but I keep wanting there to be a macro that makes it easy to delegate AsyncRead and/or AsyncWrite to an inner type, either through a pin projection or by unpinning. The vast majority of AsyncRead/Write implementations I write are just boilerplate delegation. It would be cool if that could look like a derive macro, but that would also have to be aware of pin projection

**Comment by nikomatsakis** [comment]
Yes, that is what this issue is about. The difficulty of writing wrappers that work with `Poll`, the need to use `pin-project`, etc.

**Comment by jbr** [comment]
I just realized "wrap" was intentionally vague because a solution might look more like IntoAsyncRead and IntoAsyncWrite traits, and that's probably preferable to the macro 👍
